### ISSUE TITLE
New feature to apply a shader to a specific cell in a tile map

### ISSUE THREAD

**Issue by haimat** [godotengine/godot-proposals#2173]
### Describe the project you are working on
I am working on a building/strategy game, using tile maps in Godot. While building constructions on these tile maps various parts of the map will be highlighted, based on what is going on in the game. Buildings on tiles can be damaged, demolished, "occupied", etc. All these things that may happen to a cell in the tile map have to be visible, so that the player knows what is going on on the map.

### Describe the problem or limitation you are having in your project
As said our game is based on tile maps, using tile sets. Of course I am using the same tiles over and over again in various places on the map. Now I want to highlight a specific **cell** in the map,, for example if a building has been damaged or a game character does something on that tile map cell. I don't want to do everything with multiple sprites and animation, some of these things should be made visible via shaders. The problem now is that I am not able to apply a shader only to a certain **cell** in the tile map, but the whole texture in tile set. That way I could apply my shader to a specific tile, but that would make it visible everywhere on the map where that tile is in use. We need to limit the shader effect to a specific tile used in a specific cell.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
If one would be able to apply a shader to a cell only instead of every cells where a tile is in use, then I could work much more flexible when it comes to visual effects on the tile map. For now I need to workaournd that problem by temporarily removing a tile from the map, "fake" that tile via a sprite and apply the shader to that sprite only. That is possible but makes working with tile maps and shaders much more cumbersome and complex (i.e. more code, more bugs, etc.).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Unfortunately I don't know how to implement such a feature in Godot, since I don't know enough about the inner workings of the enigne. I hope that this feature would be quite easy to implement for a seasoned Godot developer.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Not that I am aware of. As described above, the only workaround I currently know of is to (temporarily) replace the tile in a cell with a single sprite and apply the shader to that sprite. However, that means I manually have to keep track of all these sprites "in action", remove old ones and replace them again with the original tile in the map. and so on. Moreover, my current workaround also has some negative implications to the overall rendering performance, since many such sprites with their shaders can be visible and "in action" at a time.

### Is there a reason why this should be core and not an add-on in the asset library?
Frankly, I don't konw. It might be done via an add-on, but I would not see the benefit of this. I think that many people using tile maps would benefit from the requested feature.

**Comment by tavurth** [comment]
Assuming you are using `SCREEN_TEXTURE` and `SCREEN_TEXTURE_UV` you can probably do this already by making some UV calculations. 

Basically you want to have a `uniform vec2` which is the offset of your window, then a separate `uniform vec2` which is the cell size based on the current window.

Then you simply divide your screen-size into UV coordinates based on the size of these variables.

**Comment by haimat** [comment]
@tavurth That sounds like an interesting idea, thanks for that. I am quite new to Godot shaders, in particular to `SCREEN_TEXTURE`. Would you mind elaborating a bit on your idea? Thanks in advance!

**Comment by tavurth** [comment]
Please see here:

https://docs.godotengine.org/en/stable/tutorials/shading/screen-reading_shaders.html

**Comment by haimat** [comment]
@tavurth Hmm... Now that I am reading about it, does this really and fully solve my issue? Because from what I understand I can use this `SCREEN_TEXTURE` to read the current content of the screen section in question, in order to make use of that in my shader, right? But wouldn't that use everything that is in this cell at the moment?

In my case I have multiple tile maps on top of eacxh other, some with sem-opaque tiles and so on. So what is visible in a cell is the sum of all the tiles of each tile map. But I need to apply a shader to only one tile (map) in a cell, not all tiles of all lmaps in this cell. Can this be achieved with the approach you suggested?

**Comment by tavurth** [comment]
Shaders are plenty powerful:

A shader function which checks if the current UV cell should have processing should be easy enough. 

You color that cell a specific color `vec4(1., 0., 1., 0.)` and only run shader code if the color is within a specific range.

You could pass in a second texture with your designated shading locations, and then only render the shader if you find some value in that second texture. (would look like a checkerboard)

**Comment by Zireael07** [comment]
@tavurth: You are misunderstanding. SCREEN_TEXTURE grabs the screen after all tilemaps have been displayed, so he can't apply the shader to only one of the tilemaps from what I understand of the shaders.

Also applying anything with UVs to Godot's tilemaps is very difficult, I couldn't get it done when I tried. In fact, this functionality is  a very good candidate for inclusion in core unless some shader wizard comes by and shows a simple way to get this done for both square and isometric tilemaps.

**Comment by tavurth** [comment]
> only one of the tilemaps

It seems from the title he wants to apply to one specific cell of a tileset. This would just be a post processing effect.

Maybe I'm still misunderstanding but seems straight forward. Maybe you can point out what I'm missing?

**Comment by Zireael07** [comment]
> It seems from the title he wants to apply to one specific cell of a tileset. 

He has several stacked tilemaps AND he wants to apply it to one specific cell on one of the maps. Not one tile of the set, because that applies to all cells using the tile.

**Comment by tavurth** [comment]
Yep, I understand also I think we should be clear on Godot terminology:

```
TileMap -> Holds tileset
TileSet -> Holds tiles
Cell -> Individual tiles which build up the world map (x, y, TileSetId)
```

So if he wants to apply a shader to one specific `Cell` on one of the `TileSets` then he can apply a shader material to the `TileMap`, and then toggle that shader on or off using one of the methods I described above.

**Comment by tavurth** [comment]
Yep I think I understand correctly, also I think we should be clear on Godot terminology:

```
TileMap -> Holds tileset (Container)
TileSet -> Holds tiles (Resource)
Cell -> Individual tiles which build up the world map (x, y, TileSetId)
```

So if he wants to apply a shader to one specific `Cell` on one of the `TileSets` he could write a `SCREEN_TEXTURE` using shader and then toggle that shader on or off for each rendered cell using one of the methods I described above.

I didn't realise you can add a shader directly to the `TileMap`, thats pretty useful but I'm not sure if the shader would then have access to the underlying tile using `TEXTURE`, might be something to check.

**Comment by haimat** [comment]
So to clarify: I am using multiple **tile maps**, stacked on top of each other (e.g. one tile map for floor tiles, another map for wall tiles, and so on). Now I want to apply a shader to **one tile** used in **one cell** of **one tile map**. I do **not** want to apply the shader to the whole cell, i.e. to all tiles of all tile maps stacked on that cell. I want to apply the shader only to one tile in that cell, hence a post processing approach via `SCREEN_TEXTURE` does not solve my issue. This would affect all the tiles in this cell, which is not what I want.

**Comment by tavurth** [comment]
Maybe I'm not so clear, perhaps you can look at these pictures to get a better idea:

First you have the multiple TileMaps which contain data like walls etc:

![image](https://user-images.githubusercontent.com/100964/105718516-5db3cc80-5f32-11eb-8a1b-1403209a10cd.png)


Then you have a final merged TileSet which is rendered to the screen:

![image](https://user-images.githubusercontent.com/100964/105718524-61dfea00-5f32-11eb-984b-b75e5691a873.png)

Then we have a separate activation Texture (where should the shader activate):

![image](https://user-images.githubusercontent.com/100964/105718548-699f8e80-5f32-11eb-9cdb-f71d233ed59b.png)

Finally in the shader I can use this activation texture to only apply my shader to the door cell, either with an IF statement, or by using a mix with the alpha of the activation texture:

![image](https://user-images.githubusercontent.com/100964/105719112-fc402d80-5f32-11eb-83f4-ce9c36aa6c26.png)

If for some reason you want to apply the shader to a cell which is not shown in the final merged set (because it's culled) then this technique will not work.

**Comment by haimat** [comment]
@tavurth Thanks for your detailed comment, but I am afraid there is still a missunderstanding. In your last example with the screenshots, using the `SCREEN_TEXTURE`, the shader would be applied to the cell with the door. However, please mind that because of other tile maps that write to the same cell, reading from `SCEEN_TEXTURE` would give me EVERYTHING that is rendered to this cell, i.e. the "sum" of all tile maps that have rendered to this cell. But again, I don't want the shader to be applied to everything that is visible on that cell, but only to one tile in that cell.

Think of the following example: I have two tile maps that render into the same cell. One renders the floor tiles, the other tile map renders objects on to the map, also into our example cell:

![image](https://user-images.githubusercontent.com/6633976/105721429-d7dd5480-5f24-11eb-8689-f26146fe9459.png)

Now, let's assume I want to apply the shader only to the tile rendered by the "objects" tile map (the smiley in the example abiove), but not the "floor" tile map (the green tiles in the example image above). Now, with your approach using `SCREEN_TEXTURE` and the UV limited to this cell only, how can I apply a shader only to the smiley (object) but not the green backgorund (floor)?

**Comment by tavurth** [comment]
I was able to make a simple example:

<img width="1136" alt="Screenshot 2021-01-25 at 18 05 05" src="https://user-images.githubusercontent.com/100964/105723481-e2551980-5f37-11eb-9c11-17e725fee6ca.png">

Here's the node tree:

<img width="327" alt="Screenshot 2021-01-25 at 18 05 40" src="https://user-images.githubusercontent.com/100964/105723523-f00a9f00-5f37-11eb-82d2-0d7fc3c1860f.png">

Shader was applied to the second tilemap (outside of the VIewport):

```glsl
shader_type canvas_item;

uniform sampler2D tex;

void fragment() {
	
	if (texture(tex, SCREEN_UV).a > 0.) {
		COLOR = vec4(1.0, .0, .0, 1.);
	}
	else {
		COLOR = vec4(0.0, 1.0, 0.0, 1.0);
	}
}
```

Then text was set as a new `ViewportTexture` pointing to the `Viewport`

**Comment by tavurth** [comment]
The following demonstration was achieved by:

1.  Adding a `Camera2D` as a child of the `TileMap` inside the `ViewPort` (set `current`)
2. Adding a `Camera2D` to my main NodeTree. (set `current`)
3. Syncing the position of the Viewport camera to follow the main camera.

Uploading Screen Recording 2021-01-25 at 18.26.44.mov…

**Comment by tavurth** [comment]
Seems like the main thing this proposal needs is the current `vec2(x, y)` of the tile being rendered.

**Comment by eranimo** [comment]
This would be great to have. There are a ton of people asking questions about doing this

**Comment by Calinou** [comment]
Now that there are [scene tiles](https://github.com/godotengine/godot/pull/48812) in tile maps, this need may be addressed by scene tiles. It's much easier to apply a shader on individual tiles this way.

**Comment by seannybgoode** [comment]
@Zireael07 - I just made a shader work on a tilemaplayer tile. It's difficult because the UV is mapped to the entire texture atlas, not the individual tile. Once the correction was made, the rest was simple.

@tavurth - passing instance parameters to individual shaders on tiles would be super useful, because then you could re-use the same tile, and pass parameters to make that tile do something different. Example would be a shader that mixes a land and water texture, and you want to be able to rotate that mix to orient the tiles in different ways.

My use-case was for fog of war, where I wanted one shader on all my fog of war tiles, and I wanted to rotate a gradient with an alpha per tile. Depending on which direction the tile is facing, I would need to pass instance parameters to each tile. It's not currently possible with Godot, you need to (at best), re-use tiles and apply the shader to each tile, and give it whatever parameters it needs. The way it's structured defeats the purpose of doing it with a shader rather than just putitng a gradient texture together in photoshop. 

But this all has to do with bigger problems with TileMapLayer, and exposing the tiles and their data to the programmer. It's is trying so hard to be for texture display only that it's actually a pain in the ass to work with for all but the most simplistic use-cases.
